### DISCUSSION TITLE
Safety of curl_easy_pause in callback

### DISCUSSION THREAD

**Discussion by jay** [curl/curl#12059] [General]
PR #12046 added to the documentation that curl_easy_pause could be called within callbacks:

_"Unlike most other libcurl functions, [curl_easy_pause](https://curl.se/libcurl/c/curl_easy_pause.html) can be used from within callbacks."_

The thing is if you unpause a connection in a callback then couldn't it call the write callback and flip the in-callback bit to off?

For example:
- libcurl sets `Curl_set_in_callback(data, true);` then calls user's progress function
- progress function calls curl_easy_pause to unpause
- curl_easy_pause causes libcurl to set `Curl_set_in_callback(data, true);` then calls user's write function
- write function returns and libcurl sets `Curl_set_in_callback(data, false);` even though still in user's progress function

Are there other ways this could be an issue, like callback recursion if curl_easy_pause calls back into the user's progress function?

**Comment by bagder** [comment]
This can become an  issue if the callback calls *another* libcurl function (after `curl_easy_pause()` ) that is not allowed. Which then will be allowed due to the bug you mention above.
